File: packages/domain/src/accounting.ts
import { fail } from "./errors.ts";
import { addU64, asU64, subU64 } from "./math.ts";
import { epochReward } from "./reward.ts";
import type { EpochStatus } from "./types.ts";

/**
 * Exact reward accounting for one mandate. Everything is derived from a handful of stored
 * counters and the fixed invariants below; there are no ad hoc branches.
 *
 *   earned + forfeited + unresolved == accepted           (every epoch's reward is in exactly one bucket)
 *   claimed <= earned <= accepted <= max
 *   vault == max - claimed - sponsorWithdrawn             (funding conservation)
 *   vault >= (earned - claimed) + unresolved              (the provider's possible rights are always covered)
 */
export interface AccountingState {
  readonly maxRewardRaw: bigint;
  readonly acceptedRewardRaw: bigint;
  readonly totalEpochs: number;
  readonly finalizedEpochs: number;
  readonly compliantEpochs: number;
  readonly noncompliantEpochs: number;
  readonly unavailableEpochs: number;
  readonly earnedRewardRaw: bigint;
  readonly forfeitedRewardRaw: bigint;
  readonly claimedRewardRaw: bigint;
  readonly sponsorWithdrawnRaw: bigint;
}

export function newAccounting(
  maxRewardRaw: bigint,
  acceptedRewardRaw: bigint,
  totalEpochs: number,
): AccountingState {
  asU64(maxRewardRaw);
  asU64(acceptedRewardRaw);
  if (totalEpochs <= 0) return fail("InvalidEpochLength", "zero epochs");
  if (acceptedRewardRaw > maxRewardRaw)
    return fail("InvalidBudget", "accepted reward exceeds escrow");
  return {
    maxRewardRaw,
    acceptedRewardRaw,
    totalEpochs,
    finalizedEpochs: 0,
    compliantEpochs: 0,
    noncompliantEpochs: 0,
    unavailableEpochs: 0,
    earnedRewardRaw: 0n,
    forfeitedRewardRaw: 0n,
    claimedRewardRaw: 0n,
    sponsorWithdrawnRaw: 0n,
  };
}

export const allEpochsResolved = (s: AccountingState): boolean =>
  s.finalizedEpochs === s.totalEpochs;

/** Earned by the provider and not yet claimed. */
export const claimableRaw = (s: AccountingState): bigint =>
  subU64(s.earnedRewardRaw, s.claimedRewardRaw);

/** Reward of epochs that are still undecided (pending, or unavailable inside the recovery window). */
export const unresolvedRewardRaw = (s: AccountingState): bigint =>
  subU64(subU64(s.acceptedRewardRaw, s.earnedRewardRaw), s.forfeitedRewardRaw);

/** USDC still held by the reward vault. */
export const vaultBalanceRaw = (s: AccountingState): bigint =>
  subU64(subU64(s.maxRewardRaw, s.claimedRewardRaw), s.sponsorWithdrawnRaw);

/**
 * What the sponsor may withdraw now: the award surplus (`max - accepted`) at any time after award,
 * plus every forfeited reward (`accepted - earned`) once all epochs are resolved, less what was already withdrawn.
 * Nothing owed to the provider, and nothing still undecided, is ever included.
 */
export function sponsorWithdrawableRaw(s: AccountingState): bigint {
  const surplus = subU64(s.maxRewardRaw, s.acceptedRewardRaw);
  const forfeited = allEpochsResolved(s) ? subU64(s.acceptedRewardRaw, s.earnedRewardRaw) : 0n;
  return subU64(addU64(surplus, forfeited), s.sponsorWithdrawnRaw);
}

/**
 * Record one epoch's final outcome. `alreadyFinalized` comes from the caller (the program knows via
 * the epoch's unique result account); counters alone cannot tell which epochs were finalized.
 */
export function finalizeEpoch(
  s: AccountingState,
  index: number,
  outcome: EpochStatus,
  alreadyFinalized: boolean,
): AccountingState {
  if (!Number.isSafeInteger(index) || index < 0 || index >= s.totalEpochs)
    return fail("EpochOutOfRange");
  if (alreadyFinalized || s.finalizedEpochs >= s.totalEpochs) return fail("EpochAlreadyFinalized");
  const reward = epochReward(s.acceptedRewardRaw, s.totalEpochs, index);
  const base = { ...s, finalizedEpochs: s.finalizedEpochs + 1 };
  if (outcome === "Compliant") {
    return {
      ...base,
      compliantEpochs: s.compliantEpochs + 1,
      earnedRewardRaw: addU64(s.earnedRewardRaw, reward),
    };
  }
  const forfeitedRewardRaw = addU64(s.forfeitedRewardRaw, reward);
  return outcome === "NonCompliant"
    ? { ...base, noncompliantEpochs: s.noncompliantEpochs + 1, forfeitedRewardRaw }
    : { ...base, unavailableEpochs: s.unavailableEpochs + 1, forfeitedRewardRaw };
}

export type Amount = bigint | "all";

/** Provider claim. `"all"` claims everything currently claimable. */
export function claim(
  s: AccountingState,
  amount: Amount,
): { state: AccountingState; amountRaw: bigint } {
  const available = claimableRaw(s);
  const amountRaw = amount === "all" ? available : amount;
  if (amountRaw === 0n || available === 0n) return fail("NothingToClaim");
  if (amountRaw > available) return fail("ClaimExceedsEarned");
  return { state: { ...s, claimedRewardRaw: addU64(s.claimedRewardRaw, amountRaw) }, amountRaw };
}

/** Sponsor withdrawal of surplus and forfeited funds. `"all"` withdraws everything currently available. */
export function sponsorWithdraw(
  s: AccountingState,
  amount: Amount,
): { state: AccountingState; amountRaw: bigint } {
  const available = sponsorWithdrawableRaw(s);
  const amountRaw = amount === "all" ? available : amount;
  if (amountRaw === 0n || available === 0n) return fail("NothingToWithdraw");
  if (amountRaw > available) return fail("WithdrawExceedsAvailable");
  return {
    state: { ...s, sponsorWithdrawnRaw: addU64(s.sponsorWithdrawnRaw, amountRaw) },
    amountRaw,
  };
}

/** Assert every conservation invariant. Throws `InvariantViolation` naming the first one broken. */
export function assertInvariants(s: AccountingState): void {
  const bad = (name: string): never => fail("InvariantViolation", name);
  if (s.earnedRewardRaw + s.forfeitedRewardRaw > s.acceptedRewardRaw)
    bad("earned+forfeited<=accepted");
  if (s.claimedRewardRaw > s.earnedRewardRaw) bad("claimed<=earned");
  if (s.acceptedRewardRaw > s.maxRewardRaw) bad("accepted<=max");
  if (s.claimedRewardRaw + s.sponsorWithdrawnRaw > s.maxRewardRaw) bad("outflow<=funding");
  if (s.sponsorWithdrawnRaw > sponsorWithdrawableRaw({ ...s, sponsorWithdrawnRaw: 0n }))
    bad("sponsorWithdrawn<=releasable");
  if (vaultBalanceRaw(s) < claimableRaw(s) + unresolvedRewardRaw(s))
    bad("vault covers provider rights");
  if (s.finalizedEpochs !== s.compliantEpochs + s.noncompliantEpochs + s.unavailableEpochs)
    bad("epoch counters");
}

File: crates/mandate-core/src/accounting.rs
//! Exact reward accounting for one mandate, derived from a handful of counters and fixed invariants:
//!
//! ```text
//! earned + forfeited + unresolved == accepted
//! claimed <= earned <= accepted <= max
//! vault == max - claimed - sponsor_withdrawn
//! vault >= (earned - claimed) + unresolved
//! ```
use crate::math::{add_u64, sub_u64};
use crate::reward::epoch_reward;
use crate::{MandateCoreError, Result};

#[derive(Debug, Clone, Copy, PartialEq, Eq)]
pub enum EpochOutcome {
    Compliant,
    NonCompliant,
    Unavailable,
}

#[derive(Debug, Clone, Copy, PartialEq, Eq)]
pub struct AccountingState {
    pub max_reward_raw: u64,
    pub accepted_reward_raw: u64,
    pub total_epochs: u32,
    pub finalized_epochs: u32,
    pub compliant_epochs: u32,
    pub noncompliant_epochs: u32,
    pub unavailable_epochs: u32,
    pub earned_reward_raw: u64,
    /// Sum of rewards of epochs finalized NonCompliant or Unavailable.
    pub forfeited_reward_raw: u64,
    pub claimed_reward_raw: u64,
    pub sponsor_withdrawn_raw: u64,
}

/// An amount to move, or everything currently available.
#[derive(Debug, Clone, Copy, PartialEq, Eq)]
pub enum Amount {
    All,
    Exact(u64),
}

impl AccountingState {
    pub fn new(max_reward_raw: u64, accepted_reward_raw: u64, total_epochs: u32) -> Result<Self> {
        if total_epochs == 0 {
            return Err(MandateCoreError::InvalidEpochLength);
        }
        if accepted_reward_raw > max_reward_raw {
            return Err(MandateCoreError::InvalidBudget);
        }
        Ok(Self {
            max_reward_raw,
            accepted_reward_raw,
            total_epochs,
            finalized_epochs: 0,
            compliant_epochs: 0,
            noncompliant_epochs: 0,
            unavailable_epochs: 0,
            earned_reward_raw: 0,
            forfeited_reward_raw: 0,
            claimed_reward_raw: 0,
            sponsor_withdrawn_raw: 0,
        })
    }

    pub const fn all_epochs_resolved(&self) -> bool {
        self.finalized_epochs == self.total_epochs
    }

    /// Earned by the provider and not yet claimed.
    pub fn claimable_raw(&self) -> Result<u64> {
        sub_u64(self.earned_reward_raw, self.claimed_reward_raw)
    }

    /// Reward of epochs still undecided.
    pub fn unresolved_reward_raw(&self) -> Result<u64> {
        sub_u64(
            sub_u64(self.accepted_reward_raw, self.earned_reward_raw)?,
            self.forfeited_reward_raw,
        )
    }

    /// USDC still held by the reward vault.
    pub fn vault_balance_raw(&self) -> Result<u64> {
        sub_u64(
            sub_u64(self.max_reward_raw, self.claimed_reward_raw)?,
            self.sponsor_withdrawn_raw,
        )
    }

    /// Award surplus at any time, plus every forfeited reward once all epochs are resolved,
    /// less what was already withdrawn.
    pub fn sponsor_withdrawable_raw(&self) -> Result<u64> {
        let surplus = sub_u64(self.max_reward_raw, self.accepted_reward_raw)?;
        let forfeited = if self.all_epochs_resolved() {
            sub_u64(self.accepted_reward_raw, self.earned_reward_raw)?
        } else {
            0
        };
        sub_u64(add_u64(surplus, forfeited)?, self.sponsor_withdrawn_raw)
    }

    /// Record one epoch's final outcome. `already_finalized` comes from the caller (the program
    /// knows through the epoch's unique result account); counters cannot tell which epochs are done.
    pub fn finalize_epoch(
        &self,
        index: u32,
        outcome: EpochOutcome,
        already_finalized: bool,
    ) -> Result<Self> {
        if index >= self.total_epochs {
            return Err(MandateCoreError::EpochOutOfRange);
        }
        if already_finalized || self.finalized_epochs >= self.total_epochs {
            return Err(MandateCoreError::EpochAlreadyFinalized);
        }
        let reward = epoch_reward(self.accepted_reward_raw, self.total_epochs, index)?;
        let mut next = *self;
        next.finalized_epochs = bump(self.finalized_epochs)?;
        match outcome {
            EpochOutcome::Compliant => {
                next.compliant_epochs = bump(self.compliant_epochs)?;
                next.earned_reward_raw = add_u64(self.earned_reward_raw, reward)?;
            }
            EpochOutcome::NonCompliant => {
                next.noncompliant_epochs = bump(self.noncompliant_epochs)?;
                next.forfeited_reward_raw = add_u64(self.forfeited_reward_raw, reward)?;
            }
            EpochOutcome::Unavailable => {
                next.unavailable_epochs = bump(self.unavailable_epochs)?;
                next.forfeited_reward_raw = add_u64(self.forfeited_reward_raw, reward)?;
            }
        }
        Ok(next)
    }

    /// Provider claim. Returns the new state and the amount moved.
    pub fn claim(&self, amount: Amount) -> Result<(Self, u64)> {
        let available = self.claimable_raw()?;
        let amount_raw = match amount {
            Amount::All => available,
            Amount::Exact(value) => value,
        };
        if amount_raw == 0 || available == 0 {
            return Err(MandateCoreError::NothingToClaim);
        }
        if amount_raw > available {
            return Err(MandateCoreError::ClaimExceedsEarned);
        }
        let mut next = *self;
        next.claimed_reward_raw = add_u64(self.claimed_reward_raw, amount_raw)?;
        Ok((next, amount_raw))
    }

    /// Sponsor withdrawal of surplus and forfeited funds.
    pub fn sponsor_withdraw(&self, amount: Amount) -> Result<(Self, u64)> {
        let available = self.sponsor_withdrawable_raw()?;
        let amount_raw = match amount {
            Amount::All => available,
            Amount::Exact(value) => value,
        };
        if amount_raw == 0 || available == 0 {
            return Err(MandateCoreError::NothingToWithdraw);
        }
        if amount_raw > available {
            return Err(MandateCoreError::WithdrawExceedsAvailable);
        }
        let mut next = *self;
        next.sponsor_withdrawn_raw = add_u64(self.sponsor_withdrawn_raw, amount_raw)?;
        Ok((next, amount_raw))
    }

    /// Check every conservation invariant. Any failure is `InvariantViolation`.
    pub fn assert_invariants(&self) -> Result<()> {
        let bad = || MandateCoreError::InvariantViolation;
        let accounted =
            add_u64(self.earned_reward_raw, self.forfeited_reward_raw).map_err(|_| bad())?;
        if accounted > self.accepted_reward_raw
            || self.claimed_reward_raw > self.earned_reward_raw
            || self.accepted_reward_raw > self.max_reward_raw
        {
            return Err(bad());
        }
        let outflow =
            add_u64(self.claimed_reward_raw, self.sponsor_withdrawn_raw).map_err(|_| bad())?;
        if outflow > self.max_reward_raw {
            return Err(bad());
        }
        let releasable = Self {
            sponsor_withdrawn_raw: 0,
            ..*self
        }
        .sponsor_withdrawable_raw()
        .map_err(|_| bad())?;
        if self.sponsor_withdrawn_raw > releasable {
            return Err(bad());
        }
        let owed = add_u64(
            self.claimable_raw().map_err(|_| bad())?,
            self.unresolved_reward_raw().map_err(|_| bad())?,
        )
        .map_err(|_| bad())?;
        if self.vault_balance_raw().map_err(|_| bad())? < owed {
            return Err(bad());
        }
        let counted = self
            .compliant_epochs
            .checked_add(self.noncompliant_epochs)
            .and_then(|sum| sum.checked_add(self.unavailable_epochs))
            .ok_or_else(bad)?;
        if counted != self.finalized_epochs {
            return Err(bad());
        }
        Ok(())
    }
}

fn bump(counter: u32) -> Result<u32> {
    counter
        .checked_add(1)
        .ok_or(MandateCoreError::ArithmeticOverflow)
}
